[DotNetCore] Support transitive project references

mrward opened the pull request:
Fixed bug #57990 - Transitively referenced SDK projects do not show up
in IntelliSense
https://bugzilla.xamarin.com/show_bug.cgi?id=57990

With three .NET Core projects: LibC references LibB references LibA
The types defined by LibA were not available to LibC since it was not
directly referencing LibA. .NET Core projects support transitive
project references so the types defined in LibA should be available to
LibC.

 - [x] Provide DotNetProject API to return referenced projects with alias information
 - [x] Type system now using new DotNetProject API when getting project references
 - [x] .NET Core addin returns transitive project references
 - [x] DotNetCoreProject API needs a review - Is there a better API or way to do this?
 - [ ] ~~Remove code duplication in DotNetProject. New method has code similar to the existing GetReferencedAssemblyProject.~~
 - [x] Tests to be added for new DotNetProject API
 - [ ] ~~Tests to be added for type system changes when finding project references~~ Cannot see a simple way to test this.

Therzok commented on main/src/addins/MonoDevelop.DotNetCore/MonoDevelop.DotNetCore/DotNetCoreProjectExtension.cs:
```diff
@@ -608,5 +608,39 @@ void UpdateHiddenFiles (IEnumerable<ProjectFile> files)
+		void GetProjectAliases (HashSet<string> traversedProjects, List<DotNetProjectAliases> projectAliases, ConfigurationSelector configuration)
+		{
+			foreach (var projectAlias in base.OnGetReferencedAssemblyProjectAliases (configuration)) {
+				if (traversedProjects.Contains (projectAlias.Project.ItemId))
```
`if (!traversedProjects.Add())`?

slluis commented:
In PR https://github.com/mono/monodevelop/pull/2445 @mhutch is introducing new API for getting all information about references. AssemblyReference will have an Aliases property. In fact, the list of references is obtained from an msbuild target, so maybe for .net core projects it will take into account transitive references, which would make this PR unnecessary. It should be verified.

mhutch commented:
FWIW, I did some more testing on my PR and found a blocking bug in MSBuild: https://github.com/dotnet/sdk/issues/1393

mrward commented:
OK I have updated the code to use the better API from the unify project/assembly references pull request PR #2445. The changes included from the other PR were the new DotNetProject.GetReferences method and the extra properties on the AssemblyReference.

Also added some tests.

mhutch commented:
Some of that code looks suspiciously similar to mine :)

slluis commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/DotNetProject.cs:
```diff
@@ -873,13 +873,25 @@ public Task<IEnumerable<AssemblyReference>> GetReferencedAssemblies (Configurati
+
+		public Task<List<AssemblyReference>> GetReferences (ConfigurationSelector configuration, CancellationToken token)
+		{
+			return ProjectExtension.OnGetReferences (configuration, token);
```
For correctness you should use BindTask here too. The token provided to OnGetReferences would be a combination of the `token` parameter and the token provided by BindTask, created with `CancellationTokenSource.CreateLinkedTokenSource ()`

slluis commented on main/src/addins/MonoDevelop.DotNetCore/MonoDevelop.DotNetCore/DotNetCoreProjectExtension.cs:
```diff
@@ -620,5 +620,52 @@ void UpdateHiddenFiles (IEnumerable<ProjectFile> files)
+			var traversedProjects = new HashSet<string> ();
+			traversedProjects.Add (Project.ItemId);
+
+			await GetProjectAssemblyReferences (traversedProjects, references, configuration, token);
```
Isn't OnGetReferences() supposed to return all references, not only project references?

slluis commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/AssemblyReference.cs:
```diff
@@ -32,14 +32,58 @@ namespace MonoDevelop.Projects
 		{
 			FilePath = path;
-			Aliases = aliases ?? "";
+			metadata = new Dictionary<string, string> { { "Aliases", aliases } };
```
For consistency with other API, I'd like metadata to be made available through a Metadata property of type `IReadOnlyPropertySet`, like here https://github.com/mono/monodevelop/blob/master/main/src/core/MonoDevelop.Core/MonoDevelop.Projects/TargetEvaluationResult.cs#L58.

This property set can be created like this: https://github.com/mono/monodevelop/blob/master/main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs#L1287.

slluis commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/AssemblyReference.cs:
```diff
@@ -26,20 +26,71 @@
+
+		bool MetadataIsTrue (string name)
+		{
+			return string.Equals (GetMetadata (name), "true", StringComparison.OrdinalIgnoreCase);
```
MSBuildPropertyGroupEvaluated already implements boolean conversion logic, so this would be better: `return metadata.GetValue(name, false);`

slluis commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/AssemblyReference.cs:
```diff
@@ -26,20 +26,71 @@
 
 		public FilePath FilePath { get; private set; }
-		public string Aliases { get; private set; }
+		public string Aliases => GetMetadata ("Aliases") ?? "";
```
better use of api: `metadata.GetValue ("Aliases","")`
